checkout_session.go

```diff
@@ -0,0 +1,134 @@
+package stripe
+
+import (
+	"encoding/json"
+)
+
+// CheckoutSessionDisplayItemType is the list of allowed values for the display item type.
+type CheckoutSessionDisplayItemType string
+
+// List of values that CheckoutSessionDisplayItemType can take.
+const (
+	CheckoutSessionDisplayItemTypeCustom CheckoutSessionDisplayItemType = "custom"
+	CheckoutSessionDisplayItemTypePlan   CheckoutSessionDisplayItemType = "plan"
+	CheckoutSessionDisplayItemTypeSKU    CheckoutSessionDisplayItemType = "sku"
+)
+
+// CheckoutSessionLineItemParams is the set of parameters allowed for a line item
+// on a checkout session.
+type CheckoutSessionLineItemParams struct {
+	Amount      *int64    `form:"amount"`
+	Currency    *string   `form:"currency"`
+	Description *string   `form:"description"`
+	Images      []*string `form:"images"`
+	Name        *string   `form:"name"`
+	Quantity    *int64    `form:"quantity"`
+}
+
+// CheckoutSessionPaymentIntentDataTransferDataParams is the set of parameters allowed for the
+// transfer_data hash.
+type CheckoutSessionPaymentIntentDataTransferDataParams struct {
+	Destination *string `form:"destination"`
+}
+
+// CheckoutSessionPaymentIntentDataParams is the set of parameters allowed for the
+// payment intent creation on a checkout session.
+type CheckoutSessionPaymentIntentDataParams struct {
+	ApplicationFeeAmount *int64                                              `form:"application_fee_amount"`
+	CaptureMethod        *string                                             `form:"capture_method"`
+	Description          *string                                             `form:"description"`
+	OnBehalfOf           *string                                             `form:"on_behalf_of"`
+	ReceiptEmail         *string                                             `form:"receipt_email"`
+	Shipping             *ShippingDetailsParams                              `form:"shipping"`
+	StatementDescriptor  *string                                             `form:"statement_descriptor"`
+	TransferData         *CheckoutSessionPaymentIntentDataTransferDataParams `form:"transfer_data"`
+}
+
+// CheckoutSessionSubscriptionDataItemsParams is the set of parameters allowed for one item on a
+// checkout session associated with a subscription.
+type CheckoutSessionSubscriptionDataItemsParams struct {
+	Plan     *string `form:"plan"`
+	Quantity *int64  `form:"quantity"`
+}
+
+// CheckoutSessionSubscriptionDataParams is the set of parameters allowed for the subscription
+// creation on a checkout session.
+type CheckoutSessionSubscriptionDataParams struct {
+	Items           *CheckoutSessionSubscriptionDataItemsParams `form:"items"`
+	TrialEnd        *int64                                      `form:"trial_end"`
```

Comment:
This might also need its own `Params` so that it can get a `Metadata`.

Reply:
Ah yes I forgot to ask you: does it just work? Do we explicitly go through nested `Params` the right way?

Reply:
You'd have to write a test to be 100% sure, but yeah, `Params` doesn't do anything special really, so it should work fine even if nested down on a substructure like this.

The one downside of course is that `Params` has a bunch of things on it that are not relevant to this struct (e.g. idempotency key, headers, `Stripe-Account` ...), so another alternative would be just to add `Metadata map[string]string `form:"metadata"`` directly. The downside of that of course is that you'd also have to manually add the `AddMetadata` helper to the struct if you wanted to support that.

Reply:
Woot that magically just works. Nice catch! I think it's likely easier and more explicit that way than having to do something custom that users would not expect.

Reply:
Okay, WFM.